pkg/controller/cassandradatacenter/controller/statefulset.go

```diff
@@ -183,3 +295,147 @@ func newDataVolumeClaim(dataVolumeClaimSpec corev1.PersistentVolumeClaimSpec) *c
 		Spec:       dataVolumeClaimSpec,
 	}
 }
+
+// helpers
+
+func existingPods(c client.Client, cdc *cassandraoperatorv1alpha1.CassandraDataCenter) ([]corev1.Pod, error) {
+
+	podList := corev1.PodList{}
+
+	listOps := &client.ListOptions{
+		Namespace:     cdc.Namespace,
+		LabelSelector: labels.SelectorFromSet(DataCenterLabels(cdc)),
+	}
+
+	if err := c.List(context.TODO(), listOps, &podList); err != nil {
+		return nil, err
+	}
+
+	return podList.Items, nil
+}
+
+func allPodsAreRunning(pods []corev1.Pod) (bool, []string) {
+
+	podsNotInRunningPhase := make([]corev1.Pod, 0)
+
+	for _, pod := range pods {
+		if pod.Status.Phase != "Running" {
+			podsNotInRunningPhase = append(podsNotInRunningPhase, pod)
+		}
+	}
+
+	var notRunningPodNames []string
+
+	if len(podsNotInRunningPhase) > 0 {
+
+		for _, p := range podsNotInRunningPhase {
+			notRunningPodNames = append(notRunningPodNames, p.GetObjectMeta().GetName())
+		}
+
+		return false, notRunningPodNames
+	}
+
+	return true, notRunningPodNames
+}
+
+func sidecarClients(pods []corev1.Pod) map[*corev1.Pod]*sidecar.Client {
+
+	podClients := make(map[*corev1.Pod]*sidecar.Client)
+
+	options := sidecarClientOptions
+
+	for _, pod := range pods {
+		key := pod
+		podClients[&key] = sidecar.NewSidecarClient(pod.Status.PodIP, &options)
+	}
+
+	return podClients
+}
+
+func cassandraStatuses(podClients map[*corev1.Pod]*sidecar.Client) map[*corev1.Pod]string {
```

Comment:
I am not sure if we should not make go routines write to channel instead of directly putting that into map but it does work just fine.